src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11PBECipher.java

```diff
@@ -0,0 +1,217 @@
+            }
+        } else {
+            // If the key is not a P11Key, a derivation is needed. Data for
+            // derivation has to be carried either as part of the key or params.
+            // Use SunPKCS11 PBE key derivation to obtain a P11Key.
+            PBEKeySpec pbeSpec = pbes2Params.getPBEKeySpec(
+                    blkSize, svcPbeKi.keyLen, opmode, key, params, random);
+            try {
+                P11Key.P11PBEKey p11PBEKey = P11SecretKeyFactory.derivePBEKey(
+                        token, pbeSpec, svcPbeKi);
+                // The internal Cipher service uses the token where the
+                // derived key lives so there won't be any need to re-derive
+                // and use the password. The key cannot be accessed out of this
+                // class.
+                p11PBEKey.clearPassword();
+                key = p11PBEKey;
+            } catch (InvalidKeySpecException e) {
+                throw new InvalidKeyException(e);
+            } finally {
+                pbeSpec.clearPassword();
+            }
+            params = pbes2Params.getIvSpec();
+        }
+        cipher.engineInit(opmode, key, params, random);
+    }
+
+    // see JCE spec
+    @Override
+    protected void engineInit(int opmode, Key key,
+            AlgorithmParameters params, SecureRandom random)
+                    throws InvalidKeyException,
+                    InvalidAlgorithmParameterException {
+        engineInit(opmode, key, PBEUtil.PBES2Params.getParameterSpec(params),
+                random);
+    }
+
+    // see JCE spec
+    @Override
+    protected byte[] engineUpdate(byte[] input, int inputOffset,
+            int inputLen) {
+        return cipher.engineUpdate(input, inputOffset, inputLen);
+    }
+
+    // see JCE spec
+    @Override
+    protected int engineUpdate(byte[] input, int inputOffset,
+            int inputLen, byte[] output, int outputOffset)
+                    throws ShortBufferException {
+        return cipher.engineUpdate(input, inputOffset, inputLen,
+                output, outputOffset);
+    }
+
+    // see JCE spec
+    @Override
+    protected byte[] engineDoFinal(byte[] input, int inputOffset,
+            int inputLen)
+                    throws IllegalBlockSizeException, BadPaddingException {
+        return cipher.engineDoFinal(input, inputOffset, inputLen);
+    }
+
+    // see JCE spec
+    @Override
+    protected int engineDoFinal(byte[] input, int inputOffset,
+            int inputLen, byte[] output, int outputOffset)
+                    throws ShortBufferException, IllegalBlockSizeException,
+                    BadPaddingException {
+        return cipher.engineDoFinal(input, inputOffset, inputLen, output,
+                outputOffset);
+    }
+
+    // see JCE spec
+    @Override
+    protected int engineGetKeySize(Key key)
+            throws InvalidKeyException {
+        return cipher.engineGetKeySize(key);
```

Comment:
If the key is a P11PBEKey, the underlying Cipher will call P11SecretKeyFactory::convertKey with a non-PBE service algorithm and, assuming that it's in the same token and that the key is suitable for the service, will return the same key. This case should return the key length used during derivation. If the key is a PBEKey but not a P11PBEKey, there will be a call to P11SecretKeyFactory::convertKey with a non-PBE service algorithm and the key algorithm is PBE. Looks to me that, assuming that the key algorithm is suitable for the service, there will be a key derivation to finally get the length from a derived P11PBEKey. This case should not be different than using a PBEKey in a non-PBE service. @franferrax what do you think?

Reply:
We discussed this further with @franferrax and changed the approach.

While the previous approach was working for cases such as P11PBEKey and other javax.crypto.interfaces.PBEKey keys whose information was the same as used for derivation in engineInit later, we found that some cases were not working because engineGetKeySize does not have access to parameters. In particular, com.sun.crypto.provider.PBEKey keys need parameter information for a proper derivation and there could also be javax.crypto.interfaces.PBEKey keys whose information is different than passed later by parameters.

In addition to the aforementioned problems, deriving twice (in many cases) was not the best for efficiency.

There is a strong guarantee that when PBE Cipher service initialization is successful, the key size is the same as the service key size (obtained from the P11SecretKeyFactory::keyInfo map, passing the service algorithm for the lookup). This is because all keys that can successfully land on P11 PBE Cipher service are either 1) derived keys of the same algorithm, or 2) keys to be derived in which the derived key length is taken from the service key length.

As part of this change, you will see in our next commit a code fix to hold the invariant "derived PBE Cipher keys that reach a PBE Cipher service must have the same algorithm". We missed this check before because such cases are not calling P11SecretKeyFactory::convertKey —which implements the checks—, and were forwarding the key to the underlying cipher service. Notice that the underlying cipher service does not apply the same enforcement and that the underlying cipher is AES, so it will accept both 128-bit and 256-bit AES keys. Thus, the following key would have mistakenly succeeded: a PBEWithHmacSHA512AndAES_256 key passed to a PBEWithHmacSHA512AndAES_128 PBE Cipher service. We don't want to call P11SecretKeyFactory::convertKey for the case described in P11 PBE Cipher for efficiency, we just need to check the algorithms.